evm/src/cpu/kernel/asm/rlp/decode.asm

```diff
@@ -0,0 +1,145 @@
+// bytes, so that the result can be returned as a single word on the stack.
+// As per the spec, scalars must not have leading zeros.
+global decode_rlp_scalar:
+    JUMPDEST
+    // stack: pos, retdest
+    PUSH decode_int_given_len
+    // stack: decode_int_given_len, pos, retdest
+    SWAP1
+    // stack: pos, decode_int_given_len, retdest
+    // decode_rlp_string_len will return to decode_int_given_len, at which point
+    // the stack will contain (pos', len, retdest), which are the proper args
+    // to decode_int_given_len.
+    %jump(decode_rlp_string_len)
+
+// Parse the length of an RLP list from memory.
+// Pre stack: pos, retdest
+// Post stack: pos', len
+global decode_rlp_list_len:
+    JUMPDEST
+    // stack: pos, retdest
+    DUP1
+    %mload_current(@SEGMENT_RLP_RAW)
+    // stack: first_byte, pos, retdest
+    SWAP1
+    %add_const(1) // increment pos
+    SWAP1
+    // stack: first_byte, pos', retdest
+    // If first_byte is >= 0xf7, it's a > 55 byte list, and
+    // first_byte - 0xf7 is the length of the length.
+    DUP1
+    %gt_const(0xf6) // GT is native while GE is not, so compare to 0xf6 instead
+    // stack: first_byte >= 0xf7, first_byte, pos', retdest
+    %jumpi(decode_rlp_list_len_big)
+decode_rlp_list_len_small:
+    // The list length is first_byte - 0xc0.
+    // stack: first_byte, pos', retdest
+    %sub_const(0xc0)
+    // stack: len, pos', retdest
+    %stack (len, pos, retdest) -> (retdest, pos, len)
+    JUMP
+decode_rlp_list_len_big:
+    JUMPDEST
+    // The length of the length is first_byte - 0xf7.
+    // stack: first_byte, pos', retdest
+    %sub_const(0xf7)
+    // stack: len_of_len, pos', retdest
+    SWAP1
+    // stack: pos', len_of_len, retdest
+    %jump(decode_int_given_len)
+
+// Parse an integer of the given length. It is assumed that the integer will
+// fit in a single (256-bit) word on the stack.
+// Pre stack: pos, len, retdest
+// Post stack: pos', int
+decode_int_given_len:
+    JUMPDEST
+    %stack (pos, len, retdest) -> (pos, len, pos, retdest)
+    ADD
+    // stack: end_pos, pos, retdest
+    SWAP1
+    // stack: pos, end_pos, retdest
+    PUSH 0 // initial accumulator state
+    // stack: acc, pos, end_pos, retdest
+decode_int_given_len_loop:
+    JUMPDEST
+    // stack: acc, pos, end_pos, retdest
+    DUP3
+    DUP3
+    ISZERO
+    // stack: pos == end_pos, acc, pos, end_pos, retdest
+    %jumpi(decode_int_given_len_finish)
+    // stack: acc, pos, end_pos, retdest
+    %shl_const(8)
+    // stack: acc << 8, pos, end_pos, retdest
+    DUP2
+    // stack: pos, acc << 8, pos, end_pos, retdest
+    %mload_current(@SEGMENT_RLP_RAW)
+    // stack: byte, acc << 8, pos, end_pos, retdest
+    ADD
+    // stack: acc', pos, end_pos, retdest
+    // Increment pos.
+    SWAP1
+    %add_const(1)
+    SWAP1
+    // stack: acc', pos', end_pos, retdest
+    %jump(decode_int_given_len_loop)
+decode_int_given_len_finish:
```

Comment:
I was thinking of it as the part of the `decode_int_given_len` function after the loop. Do you mean to hint that it execution won't fall through (since there's a jump before)?

Reply:
Yeah I meant to make it clear that `decode_int_given_len_finish` is really a separate function not like the situation in `decode_rlp_string_len_small`. Definitely not important though

Reply:
Hm I don't quite follow what you mean by separate function, could you explain a bit more? To make sure I understand the convention

Reply:
Hm yeah sorry it feels really unimportant now. 
So with `decode_rlp_string_len_small` we have something like
```
fnA:
  condB
  %jump(fnB)
  condC
  %jump(fnC)
fnD:
  ...
```
Here not having a newline before `fnD` seems fine to me since removing the `fnD:` line wouldn't change the behavior (assuming `fnD` is never called).

But here `decode_int_given_len_finish` is called so I just feel like adding a newline before it's declaration would improve readability. But feel free to ignore, I'm definitely nitpicking

Reply:
Ah so newline before each label, except if it's a "documentation only" label? Sounds reasonable to me 👍